Add `encrypt-string` and `decrypt-to-string` functions

derwolfe opened the pull request:
@lvh - Would you mind giving this PR a review?

lvh commented on src/fernet/core.clj:
```diff
@@ -85,8 +85,19 @@
+  [key token & options]
+  (String. (apply decrypt-token key token options)))
+
 (defn encrypt [key message]
```
It would really help to understand this if it had type hints, even if it's only a compiler optimization (that goes for the other signatures too)

lvh replied:
Keep in mind that primitive-annotated functions are extra special on the JVM:

https://www.lvh.io/posts/what-the-heck-is-a-clojurelangifnlo.html

derwolfe replied:
Thanks - I'll take a look and add the type hints. :-)

lvh commented:
Oops, I just realized why the string comment is bogus.

derwolfe commented on src/fernet/core.clj:
```diff
@@ -72,7 +72,7 @@
       (invalid-token))))
 
-(defn decrypt-token [key-material token
+(defn decrypt-token [key-material ^bytes token
```
While I think this should be type-hinted as well, this might be best done with the other functions not directly affected by the PR.

lvh commented:
If one API encodes as UTF-8, the `(String. some-bytes)` should also become `String/fromBytes` of course :)

derwolfe commented:
@lvh: whoops sorry, I didn't see that you'd already commented :-). I'll make those changes as well - thanks!

derwolfe commented on src/fernet/core.clj:
```diff
@@ -72,7 +72,7 @@
       (invalid-token))))
 
-(defn decrypt-token [key-material token
+(defn decrypt-token [^String key-material ^bytes token
```
These hints look incorrect. Use `warn-on-reflection`.

derwolfe commented:
@lvh - I've gone through and hinted core based on the warnings provided by `*warn-on-reflection*`. I believe I've addressed all of your comments, so this should be ready for a fresh review :-) Thanks!

lvh commented on src/fernet/core.clj:
```diff
@@ -6,6 +6,8 @@
            (javax.crypto.spec IvParameterSpec SecretKeySpec)
            (org.bouncycastle.jce.provider BouncyCastleProvider)))
 
+(set! *warn-on-reflection* true)
```
You can specify this in project.clj:

```
  :global-vars {*warn-on-reflection* true}
```

This has the added benefit that you can keep it on for just dev.

lvh commented on src/fernet/core.clj:
```diff
@@ -41,14 +43,17 @@
-                (new SecretKeySpec key "AES")
-                (new IvParameterSpec iv)))
-            message))
+(defn aes128cbc
```
This is an insecure (in the IND-CCA2 sense) cryptographic primitive, so it should be private. (It is insecure because it is unauthenticated, meaning that it provides neither authenticity nor integrity in an IND-CCA2 scenario.)

lvh commented on src/fernet/core.clj:
```diff
@@ -85,8 +90,28 @@
+  "Encrypt the message bytes using the provided key and return the ciphertext
+  as a string"
+  [key message]
+  (String. ^String (encrypt-message key message)))
```
If it is already a dstring, why does it need to be passed to `String`?

derwolfe replied:
Whoops. Good point.

lvh commented:
LGTM modulo comments

lvh commented:
Given that the key is a string, does that mean it's in base64?

derwolfe commented:
@lvh: correct - the key is string generated by calling `Base64/encodeBase64URLSafeString` on a` byte-array` generated by `fernet.core.secure-random`. Would you like for me to change the docstrings for the encrypt/decrypt/encrypt-string/decrypt-to-string functions to describe the key?

lvh commented:
As long as it's documented somewhere that's fine. Other than that, LGTM

lvh commented on src/fernet/core.clj:
```diff
@@ -41,14 +41,17 @@
+        mode (mode {:encrypt Cipher/ENCRYPT_MODE
+                    :decrypt Cipher/DECRYPT_MODE})
+        k (new SecretKeySpec key "AES")
+        iv (new IvParameterSpec iv)]
```
I thought `Class.` was preferred, but the Java interop page doesn't suggest either way. that said, you use `String.` below; you should probably pick a style and stick with it.

lvh commented on src/fernet/core.clj:
```diff
@@ -41,14 +41,17 @@
+        iv (new IvParameterSpec iv)]
+    (.doFinal
+     (doto cipher
+       (.init ^int mode k iv))
```
Is this necessary? That's weird. init is overloaded, but all of them take int as a first parameter; the SKS and IPS should disambiguate the call

derwolfe replied:
It was generating a reflection warning without this, I did ask on clojurians about it, and this was the recommendation.

lvh commented:
OK, LGTM :)
